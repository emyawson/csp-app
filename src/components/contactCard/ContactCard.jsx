/* eslint-disable no-nested-ternary */
/* eslint-disable operator-linebreak */
/* eslint-disable no-shadow */
import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';

// styles
import './ContactCard.scss';

import '@ingka/hyperlink/style.scss';
import '@ingka/inline-message/style.scss';

import InlineMessage from '@ingka/inline-message';
import RecommendedLabel from '../recommendedLabel/RecommendedLabel.jsx';
import informationCircle from '../../../node_modules/@ingka/ssr-icon/paths/information-circle';

import RichTextEditor from '../RichTextEditor/RichText.jsx';

// Phrase
import { phrase } from '../../utils/phrase';
import ButtonCard from './Button.jsx';
import openingHoursCMS from '../../utils/openingHoursCMS';

const ContactCard = ({
  cardTitle,
  phoneNumber,
  contactType,
  index,
  recommendedButtonExists,
  standardHours,
  holidayHours,
  cardContent,
  contactLabel,
  imgCMS,
  recommended,
  disclaimer,
  id,
  buttonLink,
}) => {
  const [viewMore, setViewMore] = useState(true);

  const hoursCMS = openingHoursCMS(standardHours);
  const isStoreOpen = () => hoursCMS.afterHoursAvailability === 'open';
  const isTelephone = () => contactType === 'telephone';
  const recommendedExist = recommended;

  // For screen readers
  const ariaLabel = () => {
    if (isStoreOpen() && isTelephone()) return `${phrase.isOpen}. ${phrase.noCharge}`;
    if (isStoreOpen() && !isTelephone()) return phrase.isOpen;
    if (!isStoreOpen() && isTelephone()) return `${phrase.isClosed}. ${phrase.noCharge}`;
    return phrase.isClosed;
  };

  const standardHoursBlock = (weekDay) => standardHours.map((element, index) => (
      <div key={index}>
        <span>{weekDay[index]}</span>
        <span key={index}>{element}</span>
      </div>
  ));
  const cardClass = () => (!imgCMS ? 'card' : 'card-with-image' + ' card');
  const viewOpeningHours = () => (
    <div className="card__all-dates-container">
      <div className="card__week-days">{standardHoursBlock(hoursCMS.weekDay)}</div>
      <div className="card__opening-hours">
        <p className="card__opening-hours-headline">{phrase.holidayHoursHeadLine}</p>
        <RichTextEditor data={holidayHours} />
      </div>
    </div>
  );

  const renderInlineMessage = () => (
    <Fragment>
      <InlineMessage
        className="card__closed"
        ssrIcon={informationCircle}
        title={hoursCMS.afterHoursInline ? hoursCMS.afterHoursInline.headline : null}
        body={hoursCMS.afterHoursInline ? hoursCMS.afterHoursInline.message : null}
      />
      <p className="card__afterHoursMessage">{hoursCMS.afterHoursMessage}</p>
    </Fragment>
  );

  return (
    <Fragment>
      <div className={cardClass()}>
        {imgCMS && <img src={imgCMS} role="presentation" className="card-with-image__picture" />}

        <div className="card__title" data-cytest="contactcard-headline">
          {recommendedExist === false && index === 0 ? <RecommendedLabel /> : null}
          <h2>{cardTitle}</h2>
        </div>

        <div className="card__content-block">
          <div className="card__description" data-cytest="contactcard-description">
            <RichTextEditor data={cardContent} />
          </div>

          <ButtonCard
            id={id}
            index={recommendedButtonExists ? 1 : index}
            buttonLink={buttonLink}
            link={isTelephone()}
            phoneNumber={phoneNumber}
            ariaLabel={ariaLabel()}
            contactType={contactType}
            buttonStyles={'card__button'}
            contactLabel={contactLabel}
          />

          {/* If the IKEA store is open */}
          {isStoreOpen() && holidayHours && (
            <p className="card__open">{hoursCMS.afterHoursMessage}</p>
          )}
          {/* If the IKEA store is closed */}
          {!isStoreOpen() && holidayHours && renderInlineMessage()}
          {disclaimer && <div className="card__disclaimer">{disclaimer}</div>}
          {/* Opening and holiday hours */}
          {holidayHours && viewMore && (
            <p className="card__view-more" aria-hidden="true" onClick={() => setViewMore(false)}>
              {phrase.openingHours}
            </p>
          )}
          {!viewMore && viewOpeningHours()}
        </div>
      </div>
      <div className="separator" />
    </Fragment>
  );
};

ContactCard.propTypes = {
  phoneNumber: PropTypes.string,
  cardTitle: PropTypes.string,
  img: PropTypes.string,
  contactType: PropTypes.string,
  contactLabel: PropTypes.string,
  index: PropTypes.number,
  recommendedButtonExists: PropTypes.bool,
  standardHours: PropTypes.array,
  holidayHours: PropTypes.object,
  cardContent: PropTypes.object,
  recommended: PropTypes.bool,
  buttonLink: PropTypes.string,
  id: PropTypes.string,
};

export default ContactCard;
