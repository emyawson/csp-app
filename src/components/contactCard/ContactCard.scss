@import '../../index';

.card__description p {
  @extend %description-copybase;
  margin-top: 16px;
}
.card__button, .card__afterHoursMessage {
  margin-top: 24px;
}
.card__disclaimer {
  margin-top: 24px;
  color: #767676;
}
.card__all-dates-container {
  margin-top: 32px;
}
.card__week-days div {
  display: flex;
  margin-right: auto;
}
.card__week-days span {
  width: 120px;
  margin-top: 8px;
}
.card__view-more {
  margin-top: 24px;
  text-decoration: underline;
  cursor: pointer;
}
.card__opening-hours-headline {
  margin-top: 40px;
  margin-bottom: 10px;
  color: #111111;
  font-weight: bold;
}
.card__open, .card__closed  {
  margin-top: 24px;
}
.card-with-image {
  display: grid;
  column-gap: 42px;
  grid-template-columns: 1fr;
  @include respond(desktop) {
    max-width: 1239px;
    grid-template-columns: auto auto;
  }
}
.card-with-image__picture {
  width:100%;
  height: auto;
  margin-top: 24px;
  min-width: 335px;
  order: 1;
  @include respond(tablet) {
    max-width: 469px;
  }
  @include respond(desktop) {
    order: 0;
    margin-top: 0px;
    max-width: 100%;
    grid-row-start: 1;
    grid-row-end: 3;
    max-width: 588px;
  }
}
.card__content-block {
  order: 2;
  @include respond(tablet) {
    max-width: 467px;
  }
  @include respond(desktop) {
    max-width: 504px;
  }
}
.card__title h2 {
  @extend %large-titlebase;
}
.separator {
  max-width: 1239px;
  margin-top: 42px;
  margin-bottom: 42px;
  border-bottom: 1px solid rgba(223, 223, 223, 1);
}