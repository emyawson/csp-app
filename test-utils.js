import React from 'react';
import { Provider } from 'react-redux';
import { createStore, combineReducers } from 'redux';
import { Router } from 'react-router-dom';
import { connectRouter } from 'connected-react-router';
import { render } from '@testing-library/react';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import { history } from './src/store';
import reducers from './src/store/reducer/index';

const createRootReducer = (reducerHistory) => combineReducers({
  router: connectRouter(reducerHistory),
  ...reducers,
});

// Setup redux persist
const persistConfig = {
  key: 'root',
  storage,
  stateReconciler: autoMergeLevel2,
};
const persistedReducer = persistReducer(
  persistConfig,
  createRootReducer(history),
);

const renderWithRouter = (ui, options = {}, initialState) => ({
  ...render(
      <Provider store={createStore(persistedReducer, initialState)}>
        <Router history={history}>{ui}</Router>
      </Provider>,
  ),
});

const AllTheProviders = () => {};
const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, renderWithRouter };
