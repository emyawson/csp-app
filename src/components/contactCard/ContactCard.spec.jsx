import React from 'react';
// import { render, screen } from "@testing-library/react";
import ContactCard from './ContactCard.jsx';
// import { renderWithRouter } from "../test-utils";
import { renderWithRouter } from '../../../test-utils';

const defaultState = {};

const renderComponent = (state = {}, props) => {
  const richTextDocument = {
    nodeType: 'document',
    data: {},
    content: [
      {
        nodeType: 'paragraph',
        data: {},
        content: [
          {
            nodeType: 'text',
            value: 'Hello',
            data: {},
            marks: [{ type: 'bold' }],
          },
          {
            nodeType: 'text',
            value: ' world!',
            data: {},
            marks: [{ type: 'italic' }],
          },
        ],
      },
    ],
  };

  const mockProps = {
    afterHoursAvailability: 'open',
    afterHoursMessage: 'Open today, 08:00-21:00',
    cardContent: richTextDocument,
  };

  const combinedState = {
    ...defaultState,
    ...state,
  };
  return renderWithRouter(<ContactCard {...mockProps} {...props} />, combinedState);
};

test('renders <ContactCard/>', () => {
  const { container } = renderComponent();
  expect(container).toMatchSnapshot();

  // const linkElement = screen.getByText(/learn react/i);
  // expect(linkElement).toBeInTheDocument();
});
