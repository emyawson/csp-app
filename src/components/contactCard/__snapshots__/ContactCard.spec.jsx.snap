// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <ContactCard/> 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card__title"
      data-cytest="contactcard-headline"
    >
      <h2 />
    </div>
    <div
      class="card__content-block"
    >
      <div
        class="card__description"
        data-cytest="contactcard-description"
      >
        <div
          class="richtext-editorstyle "
        >
          <p>
            <b>
              Hello
            </b>
            <i>
               world!
            </i>
          </p>
        </div>
      </div>
      <button
        aria-label="De klantenservice is momenteel gesloten"
        class="btn btn--secondary card__button"
        data-cytest="contactcard-button"
        type="button"
      >
        <span
          class="btn__inner"
        >
          
        </span>
      </button>
    </div>
  </div>
  <div
    class="separator"
  />
</div>
`;
